src/pudl/analysis/plant_parts_eia.py

```diff
@@ -1768,3 +1491,96 @@ def assign_record_id_eia(test_df: pd.DataFrame, plant_part_col: str = "plant_par
         )
     test_df_ids = pd.concat(dfs)
     return test_df_ids
+
+
+def match_to_single_plant_part(
+    multi_gran_df: pd.DataFrame,
+    ppl: pd.DataFrame,
+    part_name: Literal[tuple(PLANT_PARTS.keys())] = "plant_gen",
```

Comment:
I have just used `Literal[PLANT_PARTS_ORDERED]`

Reply:
When I use `Literal[PLANT_PARTS_ORDERED]` it still accepts random string values without throwing an error which seems like it kind of defeats the purpose of typing here. But maybe I'm not understanding how it's supposed to work. Or there isn't great way to enforce that values are in `PLANT_PARTS_ORDERED`.

Reply:
Oh really!? wow i hadn't actually tested that... Does `Literal[tuple(PLANT_PARTS_ORDERED)]`  work? I'll be honest I do not fully understand the quirks of type hinting and how and when it gets enforced.

Reply:
If I understand correctly, the type hints don't prevent us from using invalid values. They're for static type checking with a tool like `mypy`, and for IDE integration (tab completion, etc). At runtime you can still potentially inject something bad. (though if lots of variables are typed all over the place, these kinds of inconsistencies will often get caught in static type checking).

You also can't give `Literal` anything except honest-to-god literal values like strings and ints. It can't take functions or constants defined elsewhere.

I feel like I've read a post about how to use `Enum` in this kind of situation, to define a dictionary-like structure which works like a type. But I'm not finding it immediately. Maybe it was in one of the Arjan Codes videos?

Here's an okay [overview of Literals](https://adamj.eu/tech/2021/07/09/python-type-hints-how-to-use-typing-literal/).
And another piece looking at [using Enum values for exhaustiveness checking](https://hakibenita.com/python-mypy-exhaustive-checking).

Reply:
[This is the original post](https://stackoverflow.com/questions/64522040/typing-dynamically-create-literal-alias-from-list-of-valid-values) I was looking at that was talking about how to use Literals. I think enums are mentioned in there and the top answer seems to suggest something along the lines of what Christina suggest above.